### PR TITLE
Add error code descriptions

### LINKED ISSUE #22: Add error code descriptions
Currently, whenever the firmware enters the Fault state (in other words, begins execution of the task Fault), Kucher displays the error code and says "elaboration not available". This is not cool. We need to make it display proper error messages as documented in this table: https://kb.zubax.com/display/MAINKB/Error+codes

The firmware has several tasks:

- Idle
- Fault
- Beep
- Run
- HardwareTest
- MotorIdentification
- LowLevelManipulation

They are indexed starting from zero. Each task may fail; upon failure, it returns an 8-bit error code, where value 0 is reserved to indicate success and hence can't appear among error codes. The error coding is therefore two-dimensional; for example, the error code 2 for the task Idle means "Hardware test failed", and the same code 2 for task Hardware Test means "DC voltage error".

Kucher should contain a YAML resource file listing the textual description of each error code for each task; the file should be read once at startup and used to look-up error descriptions.

### PR DESCRIPTION
**Pull request by loicgilbert**
Closes #22

### CODE REVIEW

File: kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py [hunk 1]
```diff
@@ -22,6 +22,7 @@
 
 from .base import StatusWidgetBase
 
+import yaml
```
Comment (pavel-kirienko) [inline kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py]: Please move the third-party import above the local imports, as required by PEP8 (standard first, then third-party, then local). In this case, it should be either above or under the PyQt5 import group.

Also, is PyYAML does not seem to be listed as a dependency.

File: kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py [hunk 2]
```diff
@@ -83,7 +84,16 @@ def on_general_status_update(self, timestamp: float, s: GeneralStatusView):
         self._error_code_hex.setText(f'0x{tssr.failed_task_exit_code:02X}')
         self._error_code_bin.setText(f'0b{tssr.failed_task_exit_code:08b}')
 
-        self._error_description_display.setText('(elaboration not available)')
+        file_name = 'kucher/view/main_window/telega_control_widget/task_specific_status_widget/error_codes.yml'
```
Comment (pavel-kirienko) [inline kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py]: A hard-coded path is never a good idea. Please use a relative path like:

```python
importlib.Path(__file__).absolute().parent / 'error_codes.yaml'
```

File: kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py [hunk 3]
```diff
@@ -83,7 +84,16 @@ def on_general_status_update(self, timestamp: float, s: GeneralStatusView):
...
+        with open(file_name, 'r') as f:
+            error_codes = yaml.safe_load(f)
+
+        error_description = 'unknown error'
+        for c, d in error_codes[str(tssr.failed_task_id).split('.')[-1]].items():
+            if c == tssr.failed_task_exit_code:
+                error_description = d
```
Comment (pavel-kirienko) [inline kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py]: This code gets the job done but you should be aware that the `dict` class provides a method `get()` which can default to a predefined value if the key is not found:

```python
failed_task_name = str(tssr.failed_task_id).split('.')[-1]
error_description = error_codes[failed_task_name].get(tssr.failed_task_exit_code, 'unknown error')
```

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]

**Comment by pavel-kirienko** [comment]
@dimracer could you please have a look at the error code descriptions when you're back from California? This is not urgent.

**Comment by pavel-kirienko** [comment]
Great. One last thing to fix is to list PyYAML in requirements.txt, then we should wait for Dmitry to confirm that the descriptions are okay.

**Review by pavel-kirienko — DISMISSED** [review]

**Comment by dimracer** [comment]
It would be better to extend the error description with comments. For example, "Motor model parameters are invalid. Check the motor parameters and perform the motor identification procedure."

File: kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py [hunk 4]
```diff
@@ -83,12 +89,38 @@ def on_general_status_update(self, timestamp: float, s: GeneralStatusView):
...
+    def _make_text_display(self, tool_tip: str = ''):
+        o = QTextEdit(self)
+        o.setReadOnly(True)
+        o.setLineWrapMode(True)
+
+        font = get_monospace_font()
+        font.setItalic(True)
+        o.setFont(font)
+
+        o.setStyleSheet("border: 0px;")
```
Comment (pavel-kirienko) [inline kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py]: I would prefer to have it look like a regular bordered text box, like the line display above it but multi-line. Currently, the text is rendered like a regular label, so it's easy to mistake it for part of the GUI structure as opposed to an informational message.

File: kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py [hunk 5]
```diff
@@ -83,12 +89,38 @@ def on_general_status_update(self, timestamp: float, s: GeneralStatusView):
...
         return o
+
+    def _make_text_display(self, tool_tip: str = ''):
+        o = QTextEdit(self)
+        o.setReadOnly(True)
+        o.setLineWrapMode(True)
+
+        font = get_monospace_font()
+        font.setItalic(True)
+        o.setFont(font)
```
Comment (pavel-kirienko) [inline kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py]: Let's not make it monospace; it's not code we're displaying, but regular human-readable text.

File: kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py [hunk 6]
```diff
@@ -33,13 +37,14 @@ def __init__(self, parent: QWidget):
...
 
-        self._error_code_dec = self._make_display('Exit code in decimal')
-        self._error_code_hex = self._make_display('Same exit code in hexadecimal')
-        self._error_code_bin = self._make_display('Same exit code in binary, for extra convenience')
+        self._error_code_dec = self._make_line_display('Exit code in decimal')
+        self._error_code_hex = self._make_line_display('Same exit code in hexadecimal')
+        self._error_code_bin = self._make_line_display('Same exit code in binary, for extra convenience')
 
-        self._error_description_display = self._make_display()
+        self._error_description_display = self._make_line_display()
```
Comment (pavel-kirienko) [inline kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py]: Can we please make this one non-monospace, too?

File: kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py [hunk 7]
```diff
@@ -83,12 +89,38 @@ def on_general_status_update(self, timestamp: float, s: GeneralStatusView):
...
+        file_name = get_absolute_path('view', 'main_window', 'telega_control_widget', 'task_specific_status_widget',
+                                      f'error_codes.yml', check_existence=True)
+        with open(file_name, 'r') as f:
+            error_codes = yaml.safe_load(f)
+
+        failed_task_name = str(tssr.failed_task_id).split('.')[-1]
+
+        error = error_codes[failed_task_name].get(tssr.failed_task_exit_code, 'unknown error')
+        error_description = error.get('description', 'unknown error') if type(error) is dict else error
+        error_comment = error.get('comment', '') if type(error) is dict else ''
```
Comment (pavel-kirienko) [inline kucher/view/main_window/telega_control_widget/task_specific_status_widget/fault_status_widget.py]:
```suggestion
        error_comment = error.get('comment', '') if isinstance(error, dict) else ''
```

Applies to the line above as well. Your version breaks for subclasses.

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]

**Comment by pavel-kirienko** [comment]
@loicgilbert nudge

**Comment by pavel-kirienko** [comment]
Nvm, I see the change now. Can we please keep other text fields monospace and only make the error description non-monospace?

**Comment by pavel-kirienko** [comment]
We're now back to where we started, `_error_description_display` is still monospace. Also see my old comment about `isinstance`.

**Review by pavel-kirienko — APPROVED** [review]

**Pull request merged**